api_db.php

```diff
@@ -101,7 +101,61 @@ function SQLite3_connect($trytoreconnect)
 	{
 		$from = intval($_GET["from"]);
 		$until = intval($_GET["until"]);
-		$stmt = $db->prepare("SELECT timestamp, type, domain, client, status FROM queries WHERE timestamp >= :from AND timestamp <= :until ORDER BY timestamp ASC");
+		$dbquery = "SELECT timestamp, type, domain, client, status FROM queries WHERE timestamp >= :from AND timestamp <= :until ";
+		if(isset($_GET["types"]))
+		{
+			$types = intval($_GET["types"]);
+			$typestr = "";
+			if($types & 1) // GRAVITY
```

Comment:
Passing it as a number makes the error checking rather simple and the entire API routine much more robust against possible attack vectors (like SQL injection). Although this code should be "perfectly" robust about such attacks, we can surely discuss alternatives. We could have the selector list `1,2,3` generated by JS and passed as `GET` variable and then check in PHP if the received string contains only of numbers and comma (`[0-9,]`) if you think this is a better approach.

Reply:
It should be secure to have the information passed as a list of numbers, split into an array based on commas (`explode`), and then switch on the contents. If it is not an expected value (`"alert('hack');"`), then it will fall into the default case, which can ignore the value and go to the next one in the list. It shouldn't get injected into a SQL statement.